test/unit-tests.c

```diff
@@ -1816,6 +1817,247 @@ test_process_request_free_passed_fds(void **state UNUSED)
+
+    /* bad argsz */
+    msg.in_size = sizeof(dirty_bitmap);
+    dirty_bitmap.argsz = msg.in_size + 1;
+    ret = handle_dirty_pages(&vfu_ctx, &msg);
+    assert_int_equal(-1, ret);
+    assert_int_equal(EINVAL, errno);
+
+    /* XXX VFIO_IOMMU_DIRTY_PAGES_FLAG_START */
+    dirty_bitmap.argsz = msg.in_size;
+    patch("handle_dirty_pages_get");
+    expect_value(handle_dirty_pages_get, vfu_ctx, &vfu_ctx);
+    expect_value(handle_dirty_pages_get, iovecs, &msg.out_iovecs);
+    expect_value(handle_dirty_pages_get, nr_iovecs, &msg.nr_out_iovecs);
+    expect_value(handle_dirty_pages_get, ranges, &dirty_bitmap + 1);
+    expect_value(handle_dirty_pages_get, size, msg.in_size - sizeof(dirty_bitmap));
+    will_return(handle_dirty_pages_get, 0xdeadbeef);
+
+    ret = handle_dirty_pages(&vfu_ctx, &msg);
+    assert_int_equal((int)0xdeadbeef, ret);
+
+    /*
+     * FIXME for testing VFIO_IOMMU_DIRTY_PAGES_FLAG_START and
+     * VFIO_IOMMU_DIRTY_PAGES_FLAG_STOP we must first figure out whether the DMA
+     * controller will be tracking dirty pages.
+     */
+}
+
+/* FIXME test that exec_command calls handle_device_reset */
+
+static void
+test_state_trans_notify(UNUSED void **state)
+{
+    vfu_ctx_t vfu_ctx = { 0 };
+    uint32_t vfio_device_state = 0xcafebabe;
+    vfu_migr_state_t vfu_migr_state = 0xdeadbeef;
+
+    patch("migr_state_vfio_to_vfu");
+    expect_value(migr_state_vfio_to_vfu, vfio_device_state, vfio_device_state);
+    will_return(migr_state_vfio_to_vfu, vfu_migr_state);
+    expect_value(mock_notify_migr_state_trans_cb, vfu_ctx, &vfu_ctx);
+    expect_value(mock_notify_migr_state_trans_cb, vfu_state, vfu_migr_state);
+    will_return(mock_notify_migr_state_trans_cb, 0x8badf00d);
+    assert_int_equal((int)0x8badf00d,
+        state_trans_notify(&vfu_ctx, mock_notify_migr_state_trans_cb,
+            vfio_device_state));
+}
+
+static void
+test_migr_trans_to_valid_state(UNUSED void **state)
+{
+    vfu_ctx_t vfu_ctx = { 0 };
+    struct migration migration = {
+        .callbacks.transition = (void *)0xcafebabe
+    };
+
+    patch("state_trans_notify");
+    patch("migr_state_transition");
+
+    /*
+     * XXX check that migr_trans_to_valid_state doesn't call the client
+     * callback if told not do so.
+     */
+    expect_value(migr_state_transition, migr, &migration);
+    expect_value(migr_state_transition, state,
+                 VFIO_USER_MIGR_ITER_STATE_INITIAL);
+    assert_int_equal(0, migr_trans_to_valid_state(&vfu_ctx, &migration,
+                                                  0xdeadbeef, false));
+    assert_int_equal(0xdeadbeef, migration.info.device_state);
+
+    /*
+     * TODO we should check that other members of struct migration aren't
+     * modified. This requires making struct migration private to
+     * lib/migration.c and providing functions for accessing its members.
+     */
+
+    /*
+     * XXX check that migr_trans_to_valid_state fails if the client callback
```

Comment:
nit, why all the "XXX" ? To me that means "more work to do" or "an issue here"?

Reply:
It's a way to say we're testing something, since test multiple things in a each unit test function.

Reply:
How about "TEST" or somesuch? It confuses me every time I see it :)

Reply:
Sure, it's just that it "XXX" get's highlighted on Vim :D.